### ISSUE TITLE
would you please share hype parameters of GPUs=4 for resnet50 training with us ?

### ISSUE THREAD

**Issue by mingxiaoh** [chainer/chainermn#254]
we are trying the resnet50 multinodes training experiment now, but it seems the accuracy is unexpected low. We noticed that you have tried it on GPU before(https://chainer.org/general/2017/02/08/Performance-of-Distributed-Deep-Learning-Using-ChainerMN.html), and you also displayed the experiment setting of GPUs=32, however, we only have 4 GPUs at hand, would you please share hype parameters of GPUs=4 with us? Thanks in advance.

**Comment by keisukefukuda** [comment]
Hi @mingxiaoh 
Thanks for trying ChainerMN.
Please check the points 

Does your code achieve good accuracy with a single GPU?
If not, it's not a distributed computing issue. The code in ChainerMN or Chainer's `example` directory are just examples: they are not designed to achieve the SOTA. You should try models from `ChainerCV` instead. 

Actually, the original ResNet paper used 8 GPUs. The effect of the number of GPUs is negligible up to 16-32 GPUs. We used pretty straightforward hyperparameters up to 128 GPUs.

I'm closing the issue, but feel free to re-open it if you have further questions or issues.
(However, I'm afraid we cannot support a specific model or parameter tuning issue).

**Comment by keisukefukuda** [comment]
I'm confused about 2 things in your comments. 🤔 

First, my understanding of your experiments is like this:

||Single|Multi|
|---|--------|---------|
|CPU | OK | BAD |
|GPU | Not tried| Not tried|

So why don't you try a single GPU before trying 4 GPUs?

Second, what do you mean by "multi-node" here?
Do you mean physical machines, or "CPU cores"?
I guess training ResNet 50 on a SINGLE CPU CORE takes longer than the life of the universe.
Did you do that (I guess you didn't)?

If my guess is correct, the problem is not in "multi-node", but in the batch size. 
In other words, the problem is not the number of nodes but the number of cores. 
What is the batch size you tried (in succeeded and failed cases)?

**Comment by mingxiaoh** [comment]
@keisukefukuda thanks.
First, for single node, I mean one  physical machine, for multi-nodes, it is  multi physical machines. Intel Xeon Phi was used.
Second, your table is correct.  we can achieve good accuracy on single node(CPU used), while on multi-nodes(CPU used), we found that, googlenet_v2, googlenet_v3 and resnet50 show unexpected low(<10%) validation accuracy while alexnet and googlenet can achieve SOTA accuracy. Take resnet50 for example, test with batchsize=128 on single node achieves 73.41% accuracy, while test with batchsize=32 on 4 nodes achieves <10% accuracy. We guess there might be bug in batch normalization implementation, that is why we need to reproduce your test result of resnet50 on multi-GPUs.

**Comment by keisukefukuda** [comment]
When I said "batch size", I meant the total batch size. Please check it. It's not straightforward if BS goes over 4k.
If you do not understand what I mean, I would recommend surveying some basic papers about distributed deep learning, including the one from Facebook.

About BN, we used per-GPU BN (i.e. per-minibatch BN) in the 128-GPU experiment. It should not matter that much.

**Comment by mingxiaoh** [comment]
@keisukefukuda, according to my understanding, 
case1: train_iter = chainer.iterators.MultiprocessIterator(train, args.batchsize)  # args.batchzise is local BS, 
case2: train_iter = chainer.iterators.MultiprocessIterator(train, args.batchsize // comm.size)   # args.batchzise is global BS

 we keep the global BS and other hyper parameters unchanged, when comm.size increases from 1 to 2(or 4, 8, 16), the accuracy is unexpected low.

**Comment by keisukefukuda** [comment]
> case1: train_iter = chainer.iterators.MultiprocessIterator(train, args.batchsize) # args.batchzise is local BS,
> case2: train_iter = chainer.iterators.MultiprocessIterator(train, args.batchsize // comm.size) # args.batchzise is global BS

Yes, this is correct, although keeping the global batch size while increasing number of processes is not usual.

Then, I don't quite understand this:

>  Take resnet50 for example, test with batchsize=128 on single node achieves 73.41% accuracy, while test with batchsize=32 on 4 nodes achieves <10% accuracy.

What happened here? do you mean you kept the global batch size 128 and distributed it over 4 nodes (on >70-core Xeon Phi machine)? Did you use multiple cores on a single node? If so, how?

Could you share the full output from the program?

**Comment by mingxiaoh** [comment]
@keisukefukuda , sorry for making you confused, the correct statement is:
Take resnet50 for example, test with local batchsize=128 on single node achieves 73.41% accuracy, while test with local batchsize=32 on 4 nodes achieves <10% accuracy.

for the node, I mean physical machine(say, skylake8180 or skylake6148, or broadware), yes, there are multiple cores on a single node, we use I_MPI_PIN_DOMAIN to enable jobs on multi-cores.
take skylake8180 as an example,  the command is: mpirun -ppn 1 -n 2 -genv OMP_PROC_BIND=close -genv OMP_NUM_THREADS=56 -genv I_MPI_PIN_DOMAIN=[0xffffffffffffff] -genv I_MPI_FABRICS=tcp -f ./hostfile python  train_imagenet_resnet50_2nodes.py --arch resnet50 --batchsize 128 --loaderjob 56  --epoch 32 --train_root /lustre/dataset/imagenet/img_256_resize/ILSVRC2012_img_train --val_root /lustre/dataset/imagenet/img_256_resize/ILSVRC2012_img_val --communicator naive ~/tmp/train.txt ~/tmp/val.txt -o out_resnet50_0507

**Comment by keisukefukuda** [comment]
Although I don't understand the details of Intel MPI and Xeon-Phi specific configurations, in the case, yes, BN may matter.
As I described above, in most ResNet-50 models (including the ones under `examples/`, or ChainerCV's ResNet model use normal BN layer, which is a micro-batch-wise BN. 
If you want to focus on keeping the global batch size, you need inter-process BN. ChainerMN also supports it. It is a relatively advanced feature so please refer to the code.

**Comment by mingxiaoh** [comment]
@keisukefukuda For the experiment you did for resnet50 training on multi-GPUs(https://chainer.org/general/2017/02/08/Performance-of-Distributed-Deep-Learning-Using-ChainerMN.html), did you use inter-process BN or micro-batch-wise BN? In chainercv, I see that you do not train resnet50 chainer model from scratch but covert the pretrained caffe model to chainer model directly.

**Comment by mingxiaoh** [comment]
@keisukefukuda  Does "inter-process BN" mean doing BN based on global BS while "micro-batch-wise BN" mean doing BN based on local BS? would you please tell me how to enable the "inter-process BN" feature in chainermn code?

**Comment by keisukefukuda** [comment]
Yes, you are right. 
Here's a function to do it.
https://github.com/chainer/chainermn/blob/master/chainermn/functions/batch_normalization.py#L27
It is less documented, but I guess it's pretty straightforward to use it.

However, I don't think BN is the reason, because `32` is large enough for a micro-BS. There should be another reason, but I cannot answer any more about this because you didn't show any code or program output.

**Comment by mingxiaoh** [comment]
@keisukefukuda I am not sure whether it maters or not if we use local BS for train_iter but always use global BS for val_iter.  I will try case2 and see what happen and then go back to you, thanks.
case1:
    train_iter = chainer.iterators.MultiprocessIterator(
        train, args.batchsize // comm.size, n_processes=args.loaderjob) # args.batchsize is global BS
    val_iter = chainer.iterators.MultiprocessIterator(
        val, args.val_batchsize, repeat=False, n_processes=args.loaderjob) # args.val_batchsize is global val BS and fixed to 250

case2:
    train_iter = chainer.iterators.MultiprocessIterator(
        train, args.batchsize // comm.size, n_processes=args.loaderjob) # args.batchsize is global BS
    val_iter = chainer.iterators.MultiprocessIterator(
        val, args.val_batchsize // comm.size, repeat=False, n_processes=args.loaderjob) # args.val_batchsize is global val BS and fixed to 250

**Comment by mingxiaoh** [comment]
@keisukefukuda seems that it doesn't matter whether using global BS or local BS for val_iter.
would you please kindly help check the resnet50 training code we used: https://github.com/mingxiaoh/chainermn/blob/master/examples/imagenet/train_imagenet_resnet50.py 
Thanks in advance.

**Comment by mingxiaoh** [comment]
@keisukefukuda, thanks, we use the Polynomial learning rate policy for resnet50 training, which has been PR to chainer, see 
https://github.com/chainer/chainer/pull/4693.

To run test on CPU, please install ideep4py by: 
pip install ideep4py
export CHAINER_LAZY_GRAD_SUM=1                    
export CHAINER_USE_IDEEP="auto"

**Comment by keisukefukuda** [comment]
OK, but I don't have an access to Intel Xeon Phi machines so I will test it only on GPUs. If it works well, at least it helps to narrow the problem. Give me a few days 👍

**Comment by mingxiaoh** [comment]
@keisukefukuda thanks.

**Comment by mingxiaoh** [comment]
@keisukefukuda hello, any progress for this issue so far?

**Comment by keisukefukuda** [comment]
Ah, I've been pretty busy and there's no progress yet. I think I can spare some time within this week.
Did you try GPU version as you mentioned before (with 1 and 4 GPUs)?

**Comment by mingxiaoh** [comment]
@keisukefukuda we tried it on GPU but code hangs up in "multiprocessing.set_start_method('forkserver')"

**Comment by mingxiaoh** [comment]
@keisukefukuda hello, any progress for this issue so far?

**Comment by mingxiaoh** [comment]
@keisukefukuda @niboshi 
we might have found out where the problem is.
We tried the numpy on CPU case(with iDeep disabled), replace chainer-batchmormalization by chainermn-batchmormalization, multi-node validation trend turns better.
so, here is the question, when you do multi-node test for resnet50 on your site, in the resnet50 model, did you use chainer-batchmormalization or chainermn-batchmormalization?

Case1: iDeep disabled, numpy used, resnet50 multi-node test. (chainer-batchnormalization used)

==========================================
Num process (COMM_WORLD): 4
Using naive communicator
Using resnet50 arch
Num epoch: 32
Minibatch-size: 128
Microbatch-size: 32
GPU : False
==========================================
Rank 0 [mlt-skx122]: comm.intra_rank = 0
Rank 1 [mlt-skx123]: comm.intra_rank = 0
Rank 2 [mlt-skx124]: comm.intra_rank = 0
Rank 3 [mlt-skx125]: comm.intra_rank = 0
epoch       iteration   main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  lr
1           102         4.08077     2.28204               0.11152        0.155914                  0.1
2           204         2.24751     2.2793                0.164828       0.104839                  0.1
3           305         2.11603     2.62665               0.221844       0.0954301                 0.1
4           407         1.88559     2.4713                0.341299       0.102151                  0.1
5           508         1.74191     2.44349               0.435025       0.211022                  0.1
6           610         1.40304     2.5958                0.574142       0.217742                  0.1
7           711         1.12551     4.11916               0.670483       0.130376                  0.1
8           813         0.998353    2.84913               0.711703       0.28629                   0.1
9           915         0.838145    4.24454               0.746017       0.137097                  0.1
10          1016        0.768765    3.88907               0.750928       0.223118                  0.1
     total [################..................................] 32.22%
this epoch [###############...................................] 30.89%
      1047 iter, 10 epoch / 32 epochs
  0.037786 iters/sec. Estimated time to finish: 16:11:41.768451.



Case2: iDeep disabled, numpy used, resnet50 multi-node test. (chainermn-batchnormalization used)
==========================================
Num process (COMM_WORLD): 4
Using naive communicator
Using resnet50 arch
Num epoch: 32
Minibatch-size: 128
Microbatch-size: 32
GPU : False
==========================================
Rank 0 [mlt-skx126]: comm.intra_rank = 0
Rank 1 [mlt-skx127]: comm.intra_rank = 0
Rank 2 [mlt-skx128]: comm.intra_rank = 0
Rank 3 [mlt-skx129]: comm.intra_rank = 0
epoch       iteration   main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  lr
1           102         3.16048     1.89051               0.0925245      0.229839                  0.1
2           204         1.84326     1.9956                0.270221       0.166667                  0.1
3           305         1.59036     1.91114               0.370668       0.272849                  0.1
4           407         1.4189      1.8636                0.459252       0.194892                  0.1
5           508         1.34081     2.06047               0.451423       0.379032                  0.1
6           610         1.29644     1.57334               0.483762       0.318548                  0.1
7           711         1.25448     1.92905               0.512686       0.294355                  0.1
8           813         1.18544     1.62969               0.55576        0.419355                  0.1
9           915         1.09009     1.34741               0.629596       0.544355                  0.1
